src/compiler/program.ts

```diff
@@ -748,7 +748,21 @@ namespace ts {
         }
 
         if (!tryReuseStructureFromOldProgram()) {
-            forEach(rootNames, name => processRootFile(name, /*isDefaultLib*/ false));
+            let programSize = 0;
+            for (const name of rootNames) {
+                const path = toPath(name, currentDirectory, getCanonicalFileName);
+                if (programSize <= maxProgramSize) {
+                    processRootFile(name, /*isDefaultLib*/ false);
+                    if (!hasTypeScriptFileExtension(name) && filesByName.get(path)) {
+                        programSize += filesByName.get(path).text.length;

```

Comment:
Looking in `findSourceFile` (https://github.com/Microsoft/TypeScript/blob/master/src/compiler/program.ts#L1466) I believe the filename can be present even if no file was found/read, thus you might want to check the `text` value is present before trying to access its `length`.